server/tserver/src/main/java/org/apache/accumulo/tserver/BusiestTracker.java

```diff
@@ -0,0 +1,142 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.accumulo.tserver;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+
+import org.apache.accumulo.core.data.impl.KeyExtent;
+import org.apache.accumulo.core.util.Pair;
+import org.apache.accumulo.tserver.tablet.Tablet;
+
+/**
+ * <P>
+ * Computes the N tablets that have the highest deltas for a given monotonically increasing counter.
+ * 
+ * <P>
+ * This class was parameterized to facilitate writing unit test. Writing unit test with the Tablet
+ * type would cumbersome.
+ * 
+ * @param <K>
+ *          Represents a key extent
+ * @param <T>
+ *          Represents a Tablet
+ */
+public abstract class BusiestTracker<K extends Comparable<K>,T> {
+
+  private Map<K,Long> lastCounts = Collections.emptyMap();
+  private final int numBusiestTabletsToLog;
+
+  BusiestTracker(int numBusiestTabletsToLog) {
+    this.numBusiestTabletsToLog = numBusiestTabletsToLog;
+  }
+
+  protected abstract K extractKey(T tablet);
+
+  protected abstract long extractCount(T tablet);
+
+  public List<Pair<K,Long>> computeBusiest(Collection<T> tablets) {
+
+    Comparator<Pair<K,Long>> busiestTabletComparator = new Comparator<Pair<K,Long>>() {
+      @Override
+      public int compare(Pair<K,Long> first, Pair<K,Long> second) {
+        int cmp = first.getSecond().compareTo(second.getSecond());
+        if (cmp == 0) {
+          cmp = first.getFirst().compareTo(second.getFirst());
+        }
+        return cmp;
+      }
+    };
+
+    HashMap<K,Long> counts = new HashMap<>();
+
+    PriorityQueue<Pair<K,Long>> busiestTabletsQueue =
+        new PriorityQueue<>(numBusiestTabletsToLog, busiestTabletComparator);
+
+    for (T tablet : tablets) {
+      K extent = extractKey(tablet);
+
+      // only get the count once to ensure consistency in the case of multiple threads
+      long count = extractCount(tablet);
+
+      if (count == 0)
+        continue;
+
+      counts.put(extent, count);
+
+      Long lastCount = lastCounts.get(extent);
+      if (lastCount == null)
+        lastCount = 0L;
+
+      long delta = count - lastCount;
+
+      // handle case where tablet leaves tserver and comes back OR tablet has had no activity
+      if (delta <= 0)
+        continue;
+
+      if (busiestTabletsQueue.size() < numBusiestTabletsToLog
+          || busiestTabletsQueue.peek().getSecond() < delta) {
+        if (busiestTabletsQueue.size() == numBusiestTabletsToLog) {
+          busiestTabletsQueue.remove();
```

Comment:
Could create a `LimitNPriorityQueue` subclass or some other `TopN` data structure to hold this logic, and move this logic into `add`, since it's probably the most important logic for how we're using the `PriorityQueue`. It'd be nice if it was baked in. Would make this code easier to look at (and easier to test/verify this conditional logic).

Also, it'd be nice if the above code used the same comparator... to handle the case where the last item had the same delta as the candidate to be added. Right now, it has preference for the first one entered, but when sorting later, it has a preference based on the `KeyExtent.compareTo`.

Reply:
Used `Ordering.natural().greatestOf(...)` from Guava.  This greatly simplifies the code, however it has a disadvantage over the previous code in that all tablets with a delta must be copied into a list. However it still shares the advantage of the previous code in that it does not sort all of the data.